hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -346,6 +352,12 @@ case class HoodieFileIndex(spark: SparkSession,
       Option.empty
     } else if (recordKeys.nonEmpty) {
       Option.apply(recordLevelIndex.getCandidateFiles(getAllFiles(), recordKeys))
+    } else if (recordKeys.nonEmpty && partitionStatsIndex.isIndexAvailable && !queryFilters.isEmpty) {
+      val prunedFileNames = getPrunedFileNames(prunedPartitionsAndFileSlices)
+      val shouldReadInMemory = partitionStatsIndex.shouldReadInMemory(this, queryReferencedColumns)
```

Comment:
Can you please point me to the piece of code/line numbers where we identify a referenced column as a partitioned column? (on read side) 
and how `.partitionBy("a,b,c")` turns into partition stats on the metadata write side?

That part is a still unclear to me.

Reply:
First thing I want to clarify is that partition stats are collected only when column stats is enabled, and for only those columns for which column stats is enabled.

Write path: When user will do `.partitionBy("a,b,c")`, then the logic is similar to column stats. We use the commit metadata  and convert that to partition stats. This happens in `HoodieTableMetadataUtil.convertMetadataToPartitionStatsRecords`. The difference from column stats is that the stats are aggregated by partition value in `BaseFileUtils.getColumnRangeInPartition`.

Read path: `queryReferencedColumns` here contain data filters. Partition pruning based on partition filters has already happened one level above.